Batteries/Data/List/Lemmas.lean

```diff
@@ -330,43 +330,95 @@ theorem disjoint_take_drop : ∀ {l : List α}, l.Nodup → m ≤ n → Disjoint
       refine ⟨fun h => h₀ _ (mem_of_mem_drop h) rfl, ?_⟩
       exact disjoint_take_drop h₁ (Nat.le_of_succ_le_succ h)
 
-/-! ### Chain -/
+/-! ### Pairwise -/
 
-attribute [simp, grind ←] Chain.nil
+attribute [simp, grind ←] Pairwise.nil
 
-@[simp, grind =]
-theorem chain_cons {a b : α} {l : List α} : Chain R a (b :: l) ↔ R a b ∧ Chain R b l :=
-  ⟨fun p => by cases p with | cons n p => exact ⟨n, p⟩,
-   fun ⟨n, p⟩ => p.cons n⟩
+protected theorem Pairwise.isChain (p : Pairwise R l) : IsChain R l := by
+  induction p with
+  | nil => grind
+  | cons _ l => cases l with grind
 
-theorem rel_of_chain_cons {a b : α} {l : List α} (p : Chain R a (b :: l)) : R a b :=
-  (chain_cons.1 p).1
+theorem pairwise_cons_cons :
+    Pairwise R (a :: b :: l) ↔ R a b ∧ Pairwise R (a :: l) ∧ Pairwise R (b :: l) := by
+  simp only [pairwise_cons, mem_cons, forall_eq_or_imp]
+  exact ⟨fun h => ⟨h.1.1, ⟨h.1.2, h.2.2⟩, h.2⟩, fun h => ⟨⟨h.1, h.2.1.1⟩, h.2.2⟩⟩
 
-theorem chain_of_chain_cons {a b : α} {l : List α} (p : Chain R a (b :: l)) : Chain R b l :=
-  (chain_cons.1 p).2
+/-! ### IsChain -/
 
-theorem Chain.imp' {R S : α → α → Prop} (HRS : ∀ ⦃a b⦄, R a b → S a b) {a b : α}
-    (Hab : ∀ ⦃c⦄, R a c → S b c) {l : List α} (p : Chain R a l) : Chain S b l := by
-  induction p generalizing b with grind
+attribute [simp, grind ←] IsChain.nil
+attribute [simp, grind ←] IsChain.singleton
 
-theorem Chain.imp {R S : α → α → Prop} (H : ∀ a b, R a b → S a b) {a : α} {l : List α}
-    (p : Chain R a l) : Chain S a l :=
-  p.imp' H (H a)
+@[deprecated (since := "2025-09-19")]
+alias chain_cons := isChain_cons_cons
 
-protected theorem Pairwise.chain (p : Pairwise R (a :: l)) : Chain R a l := by
-  let ⟨r, p'⟩ := pairwise_cons.1 p; clear p
-  induction p' generalizing a with grind
+theorem rel_of_isChain_cons_cons {a b : α} {l : List α} (p : IsChain R (a :: b :: l)) : R a b :=
+  (isChain_cons_cons.1 p).1
+
+@[deprecated (since := "2025-09-19")]
+alias rel_of_chain_cons := rel_of_isChain_cons_cons
+
+theorem isChain_of_isChain_cons_cons {a b : α} {l : List α} (p : IsChain R (a :: b :: l)) :
+    IsChain R (b :: l) := (isChain_cons_cons.1 p).2
+
+@[deprecated (since := "2025-09-19")]
+alias chain_of_chain_cons := isChain_of_isChain_cons_cons
+
+theorem IsChain.imp {S : α → α → Prop} (H : ∀ ⦃a b : α⦄, R a b → S a b) {l : List α}
+    (p : IsChain R l) : IsChain S l := by induction p with grind
+
+theorem IsChain.imp' {S : α → α → Prop} (HRS : ∀ ⦃a b : α⦄, R a b → S a b)
+    (Hab : ∀ ⦃c⦄, R a c → S b c) {l : List α} : IsChain R (a :: l) → IsChain S (b :: l) := by
+  cases l with grind [IsChain.imp]
```

Comment:
This theorem is a bit too niche for my taste. I don't mind keeping it and I'm sure it was useful somewhere at some time, but I couldn't find any current uses. I propose we deprecate `Chain.imp'` without a drop-in replacement. If CI finds a subtle use case then we can reevaluate before merging.

Reply:
Here is a more useful lemma that fills in the gap:
```lean
theorem IsChain.change_head (h : ∀ c, R a c → R b c) : 
    IsChain R (a :: l) → IsChain R (b :: l) := by
  cases l <;> grind
```

Reply:
Where?

Reply:
I was hoping to get rid of `imp` but I think we're stuck with it for the moment.